Code/Framework/AzCore/AzCore/Debug/Budget.h

```diff
@@ -0,0 +1,101 @@
+        }
+
+        uint32_t Crc() const
+        {
+            return m_crc;
+        }
+
+    private:
+        const char* m_name;
+        uint32_t m_crc;
+        struct BudgetImpl* m_impl = nullptr;
+    };
+} // namespace AZ::Debug
+#pragma warning(pop)
+
+// Budgets are registered and retrieved using the proxy type specialization of Budget::Get<T>. The type itself has no declaration/definition
+// other than this forward type pointer declaration
+#define AZ_BUDGET_PROXY_TYPE(name) class AzBudget##name*
+
+// Usage example:
+// In a single C++ source file:
+// AZ_DEFINE_BUDGET(AzCore);
+//
+// Anywhere the budget is used, the budget must be declared (either in a header or in the source file itself)
+// AZ_DECLARE_BUDGET(AzCore);
+//
+// The budget is usable in the same file it was defined without needing an additional declaration
+#define AZ_DEFINE_BUDGET(name)                                                                                                             \
+    template<>                                                                                                                             \
+    ::AZ::Debug::Budget* ::AZ::Debug::Budget::Get<AZ_BUDGET_PROXY_TYPE(name)>()                                                            \
+    {                                                                                                                                      \
+        static ::AZStd::atomic<::AZ::Debug::Budget*> budget;                                                                               \
+        ::AZ::Debug::Budget* out = budget.load(AZStd::memory_order_acquire);                                                               \
+        if (out)                                                                                                                           \
+        {                                                                                                                                  \
+            return out;                                                                                                                    \
+        }                                                                                                                                  \
+        else                                                                                                                               \
+        {                                                                                                                                  \
+            budget.store(&::AZ::Debug::BudgetTracker::GetBudgetFromEnvironment(#name), AZStd::memory_order_release);                       \
+            return budget;                                                                                                                 \
+        }                                                                                                                                  \
```

Comment:
Say we arrive at this code while the pointer to `budget` is being written, and we issue a store. I believe on AArch64 specifically, this already poses a problem because the write to the 64-bit pointer value may be torn.

Reply:
This would be negated by the fact that statics need to be initialized in a thread safe manner I think. Take a look at the following: https://godbolt.org/z/cx75fnd6G or the discussion here: https://stackoverflow.com/questions/8102125/is-local-static-variable-initialization-thread-safe-in-c11

Reply:
The static is initialized to nullptr.